File: course_3/module_3/programming_assignment_3/solution_3.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

// PathGraphMWIS represents the Maximum Weight Independent Set (MWIS) solver for a path graph.
type PathGraphMWIS struct {
	weights  []int // Weights of the vertices in the path graph.
	n        int   // Number of vertices.
	dp       []int // DP table for storing maximum weights for subproblems.
	solution []int // List of selected vertices in the MWIS (0-based indices).
}

// NewPathGraphMWIS initializes a new MWIS solver with the given vertex weights.
func NewPathGraphMWIS(weights []int) *PathGraphMWIS {
	n := len(weights)
	return &PathGraphMWIS{
		weights:  weights,
		n:        n,
		dp:       make([]int, n+1),
		solution: []int{},
	}
}

// ComputeMWIS computes the maximum weight independent set using dynamic programming.
func (mwis *PathGraphMWIS) ComputeMWIS() {
	if mwis.n == 0 {
		return // No vertices in the graph.
	}
	if mwis.n >= 1 {
		mwis.dp[1] = mwis.weights[0] // Base case for the first vertex.
	}

	// Fill the DP table using the recurrence relation:
	// dp[i] = max(dp[i-1], dp[i-2] + weights[i-1])
	for i := 2; i <= mwis.n; i++ {
		mwis.dp[i] = max(mwis.dp[i-1], mwis.dp[i-2]+mwis.weights[i-1])
	}
}

// ReconstructSolution reconstructs the MWIS from the computed DP table.
func (mwis *PathGraphMWIS) ReconstructSolution() {
	i := mwis.n
	for i >= 1 {
		if i == 1 || mwis.dp[i] != mwis.dp[i-1] {
			// Include the current vertex in the solution.
			mwis.solution = append([]int{i - 1}, mwis.solution...) // Add vertex to the front.
			i -= 2                                                 // Skip the adjacent vertex.
		} else {
			// Exclude the current vertex.
			i -= 1
		}
	}
}

// GetSolutionString generates a binary solution string for the specified vertices.
// Each bit represents whether the corresponding vertex is included in the MWIS.
func (mwis *PathGraphMWIS) GetSolutionString(vertices []int) string {
	// Convert the solution into a set for quick lookup.
	solutionSet := make(map[int]bool)
	for _, v := range mwis.solution {
		solutionSet[v] = true
	}

	// Generate the binary string based on the inclusion of vertices in the solution.
	var result strings.Builder
	for _, v := range vertices {
		if solutionSet[v-1] { // Convert 1-based to 0-based index.
			result.WriteString("1")
		} else {
			result.WriteString("0")
		}
	}
	return result.String()
}

// Solve solves the MWIS problem and generates the binary solution string for the specified vertices.
func (mwis *PathGraphMWIS) Solve(vertices []int) string {
	mwis.ComputeMWIS()         // Compute the DP table.
	mwis.ReconstructSolution() // Reconstruct the MWIS.
	return mwis.GetSolutionString(vertices)
}

// Utility function to compute the maximum of two integers.
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// Main function demonstrating the usage of the PathGraphMWIS class.
func main() {
	// Read the input file.
	data, err := ioutil.ReadFile("course_3/module_3/programming_assignment_3/mwis.txt")
	if err != nil {
		fmt.Println("Error reading file:", err)
		return
	}

	// Parse the weights from the file.
	lines := strings.Split(strings.TrimSpace(string(data)), "\n")
	numVertices, _ := strconv.Atoi(lines[0])
	weights := make([]int, numVertices)
	for i := 1; i <= numVertices; i++ {
		weights[i-1], _ = strconv.Atoi(lines[i])
	}

	// Initialize the MWIS solver.
	mwisSolver := NewPathGraphMWIS(weights)

	// Define the vertices to check (1-based indices).
	verticesToCheck := []int{1, 2, 3, 4, 17, 117, 517, 997}

	// Solve the problem and get the solution string.
	solutionString := mwisSolver.Solve(verticesToCheck)

	// Print the result.
	fmt.Println("Solution String:", solutionString)
}

File: course_3/module_3/programming_assignment_3/solution_3.py
class PathGraphMWIS:
    """
    Class to solve the Maximum Weight Independent Set (MWIS) problem in a path graph.
    """

    def __init__(self, weights):
        """
        Initialize the graph with the vertex weights.
        :param weights: List of weights for the vertices in the path graph.
        """
        self.weights = weights
        self.n = len(weights)
        self.dp = [0] * (self.n + 1)  # DP table for storing max weights.
        self.solution = []  # List to store the indices of selected vertices.

    def compute_mwis(self):
        """
        Compute the maximum weight independent set using dynamic programming.
        """
        if self.n == 0:
            return
        if self.n >= 1:
            self.dp[1] = self.weights[0]

        # Fill the DP table.
        for i in range(2, self.n + 1):
            self.dp[i] = max(self.dp[i - 1], self.dp[i - 2] + self.weights[i - 1])

    def reconstruct_solution(self):
        """
        Reconstruct the MWIS from the computed DP table.
        """
        i = self.n
        while i >= 1:
            if i == 1 or self.dp[i] != self.dp[i - 1]:  # Current vertex is included.
                self.solution.append(i - 1)  # Convert to 0-based index.
                i -= 2
            else:  # Current vertex is excluded.
                i -= 1

        self.solution.reverse()  # Reverse to get the solution in ascending order.

    def get_solution_string(self, vertices):
        """
        Generate the 8-bit solution string for the specified vertices.
        :param vertices: List of vertices (1-based indices) to check.
        :return: An 8-bit string where 1 indicates inclusion in MWIS.
        """
        solution_set = set(self.solution)
        return "".join("1" if v - 1 in solution_set else "0" for v in vertices)

    def solve(self, vertices):
        """
        Solve the MWIS problem and generate the solution string for the specified vertices.
        :param vertices: List of vertices to check.
        :return: 8-bit solution string.
        """
        self.compute_mwis()
        self.reconstruct_solution()
        return self.get_solution_string(vertices)


# Example usage
if __name__ == "__main__":
    # Read weights from the file.
    with open("mwis.txt", "r") as file:
        lines = file.readlines()
        num_vertices = int(lines[0].strip())
        weights = [int(line.strip()) for line in lines[1:]]

    # Initialize the MWIS solver.
    mwis_solver = PathGraphMWIS(weights)

    # Define the vertices to check (1-based indices).
    vertices_to_check = [1, 2, 3, 4, 17, 117, 517, 997]

    # Solve the problem and get the solution string.
    solution_string = mwis_solver.solve(vertices_to_check)

    # Print the result.
    print("Solution String:", solution_string)
